### DISCUSSION TITLE
support for ATSAMD21E18A controller

### DISCUSSION THREAD

**Discussion by svsramaraju** [ARMmbed/DAPLink#1044] [General]
Hi ,
i am new to DAPLink  porting. i would like to port DAP link for " ATSAMD21E18A" processor . i have seen support for Atmel controller "sam3u2c_bl" board . from the make file i have observed that the following source files required to build the application
..\..\..\source\board\sam3u2c_bl.c  
           ..\..\..\source\daplink\HardFault_Handler.c  
           ..\..\..\source\daplink\bootloader\daplink.c  
           ..\..\..\source\daplink\bootloader\main_bootloader.c  
           ..\..\..\source\daplink\circ_buf.c  
           ..\..\..\source\daplink\cortex_m.c  
           ..\..\..\source\daplink\crc32.c  
           ..\..\..\source\daplink\daplink_debug.c  
           ..\..\..\source\daplink\drag-n-drop\file_stream.c  
           ..\..\..\source\daplink\drag-n-drop\flash_decoder.c  
           ..\..\..\source\daplink\drag-n-drop\flash_intf.c  
           ..\..\..\source\daplink\drag-n-drop\flash_manager.c  
           ..\..\..\source\daplink\drag-n-drop\iap_flash_intf.c  
           ..\..\..\source\daplink\drag-n-drop\intelhex.c  
           ..\..\..\source\daplink\drag-n-drop\vfs_manager.c  
           ..\..\..\source\daplink\drag-n-drop\vfs_user.c  
           ..\..\..\source\daplink\drag-n-drop\virtual_fs.c  
           ..\..\..\source\daplink\error.c  
           ..\..\..\source\daplink\flash_hal.c  
           ..\..\..\source\daplink\info.c  
           ..\..\..\source\daplink\sdk_stub.c  
           ..\..\..\source\daplink\settings\settings.c  
           ..\..\..\source\daplink\settings\settings_rom_stub.c  
           ..\..\..\source\daplink\util.c  
           ..\..\..\source\daplink\validation.c  
           ..\..\..\source\hic_hal\atmel\sam3u2c\flash_hal_SAM3U.c  
           ..\..\..\source\hic_hal\atmel\sam3u2c\gpio.c  
           ..\..\..\source\hic_hal\atmel\sam3u2c\read_uid.c  
           ..\..\..\source\hic_hal\atmel\sam3u2c\system_SAM3U.c  
           ..\..\..\source\hic_hal\atmel\sam3u2c\uart.c  
           ..\..\..\source\hic_hal\atmel\sam3u2c\usb_config.c  
           ..\..\..\source\hic_hal\atmel\sam3u2c\usbd_ATSAM3U2C.c  
           ..\..\..\source\rtos_none\SysTick_Handler.c  
           ..\..\..\source\rtos_none\cmsis_os2_port.c  
           ..\..\..\source\target\target_board.c  
           ..\..\..\source\target\target_default.c  
           ..\..\..\source\usb\msc\usbd_core_msc.c  
           ..\..\..\source\usb\msc\usbd_msc.c  
           ..\..\..\source\usb\usbd_core.c 

i would like to check that the following HAL files need to be  rewrite for ATSAMD21E18A"?

           ..\..\..\source\hic_hal\atmel\sam3u2c\flash_hal_SAM3U.c  
           ..\..\..\source\hic_hal\atmel\sam3u2c\gpio.c  
           ..\..\..\source\hic_hal\atmel\sam3u2c\read_uid.c  
           ..\..\..\source\hic_hal\atmel\sam3u2c\system_SAM3U.c  
           ..\..\..\source\hic_hal\atmel\sam3u2c\uart.c  
           ..\..\..\source\hic_hal\atmel\sam3u2c\usb_config.c  
           ..\..\..\source\hic_hal\atmel\sam3u2c\usbd_ATSAM3U2C.c  
can someone provide some information related to required steps for porting?

Regards,
Raju

**Comment by mbrossard** [comment]
Porting DAPLink to a new interface is quite a complex task. Potentially all files in `source/hic_hal/atmel/sam3u2c` need to be replaced by an equivalent for SAMD21. You might find a lot of them in CMSIS packs or SDKs. The hardest to replace is the `usbd_*.c` (USB device driver). Currently you would have to write a driver for the `RL-USB` stack. If you are not really confident that you can write that driver, I would advise not to move forward.

Once we add support for the TinyUSB stack it should get much easier and we are more likely to document the process.
Depending on your needs you could try adding support for SAMD21 to https://github.com/majbthrd/DapperMime (SAMD21 support already exists in TinyUSB). Another option is https://github.com/ciniml/rust-dap that supports another SAMD21 model.
